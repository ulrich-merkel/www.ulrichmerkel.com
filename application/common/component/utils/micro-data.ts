/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
