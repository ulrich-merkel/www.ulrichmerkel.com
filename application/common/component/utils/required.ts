/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
