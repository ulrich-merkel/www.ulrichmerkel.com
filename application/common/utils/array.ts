/**
 * Es6 module for handling array related utilities.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */

/**
 * Check if given "input" is a valid array with at least one entry.
 *
 * @param {Array<*>} array - The array to be checked
 * @returns {boolean} Whether this is an array with some entries
 */
export function isValidArray(array?: any[]): boolean {
    return Array.isArray(array) && array.length > 0;
}
