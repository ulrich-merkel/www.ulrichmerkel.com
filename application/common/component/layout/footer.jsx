/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Layout components merge modules to bigger parts of the
 * overall page layout.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.3
 *
 * @requires react
 * @requires prop-types
 * @requires classnames
 * @requires common/component/decorator/add-content
 * @requires common/utils/content
 * @requires common/component/grid/container
 * @requires common/component/grid/spaced
 * @requires common/component/grid/row
 * @requires common/component/grid/col
 * @requires common/component/module/menu
 * @requires common/component/element/nav
 * @requires common/component/element/button
 * @requires common/component/element/small
 *
 * @changelog
 * - 0.0.3 Moved to stateless function
 * - 0.0.2 Rewritten for es2015
 * - 0.0.1 Basic functions and structure
 */
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import addContent from './../decorator/add-content';
import { getContentSection } from './../../utils/content';
import GridContainer from './../grid/container';
import GridSpaced from './../grid/spaced';
import GridRow from './../grid/row';
import GridCol from './../grid/col';
import ModuleMenu from './../module/menu';
import Nav from './../element/nav';
import Button from './../element/button';
import Small from './../element/small';

/**
 * Function representing a component to return a single react child element.
 *
 * @param {Object} [props] - The current component props
 * @returns {React.Element} React component markup
 */
function LayoutFooter(props) {

    const {
        className,
        content,
        handleScrollTop
    } = props;

    const contentSection = getContentSection(content);
    const componentClassName = classnames('l-footer', className);

    return (
        <footer className={componentClassName} itemScope itemType='http://schema.org/WPFooter' role='contentinfo'>

            <Button
                className='l-footer__button--up c-font-icon--chevron-up'
                classNameLabel='is-visually-hidden'
                onClick={handleScrollTop}
                title={contentSection('btnTitle')}
                type='button'
                role='button'
            >
                {contentSection('btnLabel')}
            </Button>

            <Nav>
                <GridContainer>
                    <GridSpaced>
                        <GridRow>
                            <GridCol>
                                <ModuleMenu content={contentSection('menu.imprint')} />
                            </GridCol>
                        </GridRow>
                        <GridRow>
                            <GridCol>
                                <ModuleMenu content={contentSection('menu.social')} />
                            </GridCol>
                        </GridRow>
                        <GridRow>
                            <GridCol>
                                <Small>
                                    &copy; <strong>{contentSection('copy.name')}</strong> {contentSection('copy.date')}
                                </Small>
                            </GridCol>
                        </GridRow>
                    </GridSpaced>
                </GridContainer>
            </Nav>

        </footer>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 * @property {string} [className] - The component css class names - will be merged into component default classNames
 * @property {Function} [handleSearchChange=Function.prototype] - Function handling to top scrolling
 * @property {Object} [content={}] - The component content config
 */
LayoutFooter.propTypes = {
    className: PropTypes.string, // eslint-disable-line  react/require-default-props
    handleScrollTop: PropTypes.func,
    content: PropTypes.objectOf(PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.array,
        PropTypes.object
    ]))
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {Object}
 * @see LayoutFooter.propTypes
 */
LayoutFooter.defaultProps = {
    handleScrollTop: Function.prototype,
    content: {}
};

export default addContent('LayoutFooter')(LayoutFooter);
