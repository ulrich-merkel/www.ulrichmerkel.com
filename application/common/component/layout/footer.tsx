/**
 * Es6 module for React Component.
 * Layout components merge modules to bigger parts of the
 * overall page layout.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import { withRouter } from 'react-router';
import classnames from 'classnames';
import { noop } from 'lodash';

import { addContent } from '../decorator/add-content';
import { getContentSection } from '../../utils/content';
import { GridContainer } from '../grid/container';
import { GridSpaced } from '../grid/spaced';
import { GridRow } from '../grid/row';
import { GridCol } from '../grid/col';
import { ModuleMenu } from '../module/menu';
import { Nav } from '../element/nav';
import { Button } from '../element/button';
import { Small } from '../element/small';

type Props = {
    className: string;
    content: any;
    handleScrollTop: (event) => void;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @param {object} [props] - The current component props
 * @param {string} [props.className] - The component css class names - will be merged into component default classNames
 * @param {object} [props.content={}] - The component content config
 * @param {Function} [props.handleSearchChange=Function.prototype] - Function handling to top scrolling
 * @returns {ReactElement} React component markup
 */
export const LayoutFooter: FunctionComponent<Props> = (props) => {
    const { className, content, handleScrollTop = noop } = props;

    const contentSection = getContentSection(content);
    const componentClassName = classnames('l-footer', className);

    return (
        <footer
            className={componentClassName}
            itemScope
            itemType="http://schema.org/WPFooter"
            role="contentinfo"
        >
            <Button
                className="l-footer__button--up c-font-icon--chevron-up"
                classNameLabel="is-visually-hidden"
                onClick={handleScrollTop}
                title={contentSection('btnTitle')}
                type="button"
                role="button"
            >
                {contentSection('btnLabel')}
            </Button>

            <Nav>
                <GridContainer>
                    <GridSpaced>
                        <GridRow>
                            <GridCol>
                                <ModuleMenu
                                    content={contentSection('menu.imprint')}
                                />
                            </GridCol>
                        </GridRow>
                        <GridRow>
                            <GridCol>
                                <ModuleMenu
                                    content={contentSection('menu.social')}
                                />
                            </GridCol>
                        </GridRow>
                        <GridRow>
                            <GridCol>
                                <Small>
                                    &copy;{' '}
                                    <strong>
                                        {contentSection('copy.name')}
                                    </strong>{' '}
                                    {contentSection('copy.date')}
                                </Small>
                            </GridCol>
                        </GridRow>
                    </GridSpaced>
                </GridContainer>
            </Nav>
        </footer>
    );
};

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 *
 * @type {ReactElement}
 */
export const LayoutFooterConnected = withRouter(
    addContent('LayoutFooter')(LayoutFooter)
);
