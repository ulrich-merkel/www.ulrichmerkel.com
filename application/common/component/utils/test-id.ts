/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
