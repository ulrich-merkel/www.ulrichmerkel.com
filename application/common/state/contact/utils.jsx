/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 *
 * @TODO Should be refactored and moved to actions
 *
 * @requires common/utils/validation
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';

/**
 * Validation helper to check all inputs.
 *
 * @function
 * @param {object} [state={}] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
function validate(state = {}) {
    return {
        name: required(state.name),
        email: required(state.email) && email(state.email),
        subject: required(state.subject),
        message: required(state.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @function
 * @param {object} state - The current redux contact state
 * @returns {boolean} The validation status
 */
function isValid(state) {
    const validated = validate(state);

    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

function canSendForm(state) {
    const isStateValid = isValid(state);

    return isStateValid
        && state.pristine
        && !state.pending
        && isBrowser(state);
}

export {
    canSendForm,
    isValid,
    validate
};
