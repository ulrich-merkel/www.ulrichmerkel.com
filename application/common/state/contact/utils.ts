/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';
import { ContactFormType, ValidateType } from './types';

/**
 * Validation helper to check all inputs.
 *
 * @param {object} [contactForm] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
export function validate(contactForm: ContactFormType): ValidateType {
    return {
        name: required(contactForm?.name),
        email: required(contactForm?.email) && email(contactForm?.email),
        subject: required(contactForm?.subject),
        message: required(contactForm?.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} The validation status
 */
export function isValid(contactForm: ContactFormType): boolean {
    const validated = validate(contactForm);
console.log({validated, contactForm})
    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

/**
 * Check if the form can be send.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} Whether the submit button is enabled or not
 */
export function canSendForm(contactForm: ContactFormType): boolean {
    const isStateValid = isValid(contactForm);

    return (
        isStateValid &&
        contactForm.pristine &&
        !contactForm.pending &&
        isBrowser(contactForm)
    );
}
