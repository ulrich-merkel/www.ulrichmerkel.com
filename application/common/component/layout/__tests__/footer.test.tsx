import * as React from 'react';
import { render } from '../../../../__tests__/utils/test-utils';
import { LayoutFooter, LayoutFooterConnected } from '../footer';

describe('LayoutFooter', function fnDescribe() {
    const props = {
        handleScrollTop: jest.fn(),
        content: {
            footer: 'test'
        },
        className: 'layout-footer'
    };

    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <LayoutFooter {...props}>Footer Children</LayoutFooter>
        );
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render correctly with default props', function fnIt() {
        const { asFragment } = render(<LayoutFooter />);
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render correctly the connected version', function fnIt() {
        const { asFragment } = render(
            <LayoutFooterConnected {...props}>
                Footer Children
            </LayoutFooterConnected>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
