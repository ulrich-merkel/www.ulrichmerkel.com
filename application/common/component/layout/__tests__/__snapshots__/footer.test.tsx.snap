// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutFooter should render correctly 1`] = `
<DocumentFragment>
  <footer
    class="l-footer layout-footer"
    itemscope=""
    itemtype="http://schema.org/WPFooter"
    role="contentinfo"
  >
    <button
      class="c-btn l-footer__button--up c-font-icon--chevron-up"
      data-testid="button-scroll-top"
      role="button"
      title=""
      type="button"
    >
      <span
        class="c-btn__label is-visually-hidden"
      />
    </button>
    <nav
      class="m-nav"
      role="navigation"
    >
      <section
        class="l-grid"
      >
        <div
          class="l-grid__spaced--small"
        >
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            />
          </div>
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            />
          </div>
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            >
              <small
                class="c-type--small"
              >
                © 
                <strong />
                 
              </small>
            </div>
          </div>
        </div>
      </section>
    </nav>
  </footer>
</DocumentFragment>
`;

exports[`LayoutFooter should render correctly the connected version 1`] = `
<DocumentFragment>
  <footer
    class="l-footer layout-footer"
    itemscope=""
    itemtype="http://schema.org/WPFooter"
    role="contentinfo"
  >
    <button
      class="c-btn l-footer__button--up c-font-icon--chevron-up"
      data-testid="button-scroll-top"
      role="button"
      title=""
      type="button"
    >
      <span
        class="c-btn__label is-visually-hidden"
      />
    </button>
    <nav
      class="m-nav"
      role="navigation"
    >
      <section
        class="l-grid"
      >
        <div
          class="l-grid__spaced--small"
        >
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            />
          </div>
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            />
          </div>
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            >
              <small
                class="c-type--small"
              >
                © 
                <strong />
                 
              </small>
            </div>
          </div>
        </div>
      </section>
    </nav>
  </footer>
</DocumentFragment>
`;

exports[`LayoutFooter should render correctly with default props 1`] = `
<DocumentFragment>
  <footer
    class="l-footer"
    itemscope=""
    itemtype="http://schema.org/WPFooter"
    role="contentinfo"
  >
    <button
      class="c-btn l-footer__button--up c-font-icon--chevron-up"
      data-testid="button-scroll-top"
      role="button"
      title=""
      type="button"
    >
      <span
        class="c-btn__label is-visually-hidden"
      />
    </button>
    <nav
      class="m-nav"
      role="navigation"
    >
      <section
        class="l-grid"
      >
        <div
          class="l-grid__spaced--small"
        >
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            />
          </div>
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            />
          </div>
          <div
            class="l-grid__row"
          >
            <div
              class="l-grid__col--12"
            >
              <small
                class="c-type--small"
              >
                © 
                <strong />
                 
              </small>
            </div>
          </div>
        </div>
      </section>
    </nav>
  </footer>
</DocumentFragment>
`;
